In [72]:
url = r"C:\Users\princ\OneDrive\Documents\Chapter 1\Python\Machine Learning\ML Data\mail_data.csv"

Importing The Dependencies

In [73]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import cross_val_score

Data Loading and Preprocessing

In [74]:
#Loading the Data
data = pd.read_csv(url)

In [75]:
#Printing first Five rows of the data
data.head()

,Category,Message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [76]:
data.shape

(5572, 2)

In [77]:
#Checking for any missing values
data.isnull().sum()

Category    0
Message     0
dtype: int64

In [78]:
#Encoding the Category column
le = LabelEncoder()
data['Numeric_Category'] = le.fit_transform(data['Category'])

In [79]:
data.head()

,Category,Message,Numeric_Category
0,ham,"Go until jurong point, crazy.. Available only ...",0
1,ham,Ok lar... Joking wif u oni...,0
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,1
3,ham,U dun say so early hor... U c already then say...,0
4,ham,"Nah I don't think he goes to usf, he lives aro...",0


Seprating The data to the Features and Target Column

In [80]:
X = data['Message']
Y = data['Numeric_Category']

Spam Mail--1

Ham Mail--0

Feature Extraction

In [83]:
#Now converting the texual data column Message to Numerical data
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
vectorizer.fit(X)
X = vectorizer.transform(X)

In [84]:
print(X)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 74098 stored elements and shape (5572, 8709)>
  Coords	Values
  (0, 1079)	0.32645117023873077
  (0, 1313)	0.24417482890859651
  (0, 1761)	0.31163292870610654
  (0, 1763)	0.2757872678027423
  (0, 2057)	0.2757872678027423
  (0, 2334)	0.25281395947472957
  (0, 3567)	0.1478647506898016
  (0, 3611)	0.15304155020494284
  (0, 3651)	0.18033180630708256
  (0, 4110)	0.10707657674366397
  (0, 4370)	0.32645117023873077
  (0, 4497)	0.2757872678027423
  (0, 5567)	0.15636623031957597
  (0, 5954)	0.25533539230157154
  (0, 7690)	0.15550627816331294
  (0, 8080)	0.23000344108357726
  (0, 8320)	0.1824010162830269
  (0, 8544)	0.2208188335194995
  (1, 4338)	0.5236804332035243
  (1, 4533)	0.4083258549263009
  (1, 5534)	0.2718944069420321
  (1, 5563)	0.5466243141314314
  (1, 8446)	0.43162957585464123
  (2, 77)	0.23031755640837923
  (2, 401)	0.22254722246589276
  :	:
  (5570, 1790)	0.13631002689236146
  (5570, 1798)	0.28297500390378505
  (5570, 2602

In [85]:
#Loading the Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

In [86]:
models = [LogisticRegression(max_iter = 10000),DecisionTreeClassifier(max_depth = 5),SVC(kernel = 'linear',C = 1.0),KNeighborsClassifier(n_neighbors = 5,weights = 'uniform')]

In [87]:
def compare_model_cross_validation():
    for model in models:
        cv_score = cross_val_score(model,X,Y,cv = 5)
        mean_Accuracy = round((sum(cv_score)/len(cv_score))*100,2)
        print('Cross Validation Score for the ',model,'=',cv_score)
        print('Mean Accuracy of the ',model,'=',mean_Accuracy)
        print('---------------------------------------------------------------------------------')
        
        

In [88]:
compare_model_cross_validation()

Cross Validation Score for the  LogisticRegression(max_iter=10000) = [0.96053812 0.96053812 0.96140036 0.95691203 0.96319569]
Mean Accuracy of the  LogisticRegression(max_iter=10000) = 96.05
---------------------------------------------------------------------------------
Cross Validation Score for the  DecisionTreeClassifier(max_depth=5) = [0.95156951 0.95515695 0.9497307  0.94524237 0.95332136]
Mean Accuracy of the  DecisionTreeClassifier(max_depth=5) = 95.1
---------------------------------------------------------------------------------
Cross Validation Score for the  SVC(kernel='linear') = [0.9793722  0.98565022 0.98025135 0.98114901 0.98204668]
Mean Accuracy of the  SVC(kernel='linear') = 98.17
---------------------------------------------------------------------------------
Cross Validation Score for the  KNeighborsClassifier() = [0.90852018 0.90762332 0.90754039 0.95960503 0.96140036]
Mean Accuracy of the  KNeighborsClassifier() = 92.89
-----------------------------------------

So as we can see that in betwen all of these four models Support Vector Classifier got the Highest Accuracy of 98.17%